### DISCUSSION TITLE
EG4 - LL (12v/400Ah) Questions / Docs / Can test

### DISCUSSION THREAD

**Discussion by epierce78** [mr-manuel/venus-os_dbus-serialbattery#46] [General]
@mr-manuel : 

I'm excited to see the addition of the 12V EG4-LL driver.  I have 2 of these for a total of 800Ah.  I'm up and running on Venus OS 3.40 beta (I do have a GX device and not rpi) and the nightly build of your version of the driver.

I'm running into a few issues and I do have the capability to test/debug:
- I get spurious alerts on the Cerbo / in VRM for lost of comms with the BMS.  I'm watching the logs (in I believe full debug) and I'm not seeing any issue with data coming back.
- The documentation for this BMS (whatever it is) is scarce at best.  I was able to find the PDF linked in eg4_ll.py, but not at the EG4 website / link itself.  That returns a 404.  I'm wondering if there's a source from which anyone who worked on this driver used beyond those?
- I'd love to figure out single USB / multiple battery comms.  I have mine linked via cable, but of course, that doesn't really work.  Only see a single battery.  I have another cable on order and can use BatteryAggregator.  That said, with the right pointer/docs, figuring out how to send commands that address a master + other battery packs would be really nice.
- I'm seeing DVCC charge current values come back when discharging.  Also see that I have at least one cell that's fairly out of balance relative to the rest.  I'd like to see if I can fix that and if it's not implemented, tell DVCC what the max recommended charging current is.  I think it'd help with this BMS during charges, especially on 'short cycles' - mine are installed in our RV and it's not uncommon to discharge down to 90-95% only.  In that case the BMS is *really* touchy.  Too much voltage / too many amps, the BMS will alarm and go into protection.  I've had to set values far lower than what you'd typically expect to see with Lifepo4's.
- Any pointers on tools / docs that I could use to speed testing?

Ok -- that's a lot.  Appreciate any comments/insight.  Having this working (even better) will work better than the shunt we've used thus far.

Thanks.

**Comment by mr-manuel** [comment]
I requested the protocol documentation a few days back and got them today. @tuxntoast could you take a look at the protocol documentation?

[EG4-LL  - MODBUS Communication Protocol (2024.07.05).pdf](https://github.com/user-attachments/files/16112198/EG4-LL.-.MODBUS.Communication.Protocol.2024.07.05.pdf)

Next week I will merge the version for multiple batteries on one RS485 adapter with the main branch: https://github.com/mr-manuel/venus-os_dbus-serialbattery/releases/tag/v1.3.20240623modbus

**Reply by mr-manuel** [reply]
You could test with this version: https://github.com/mr-manuel/venus-os_dbus-serialbattery/releases/tag/v1.4.20240706modbus Here you can specify multiple addresses.

**Reply by epierce78** [reply]
I am working on testing and attempting to debug this.  I found a couple of small Python bugs I'll PR later, they're not stopping me.

As it stands, I've tried almost every variation I can think of, connecting the 2 RS485 ports each battery has in terms of where the USB->RS485 adapter is connected and the cable that comes with the batteries for interconnection.  I can always get address 0x01 to work as long as one battery has the DIP switches in that position.  The 0x01 address will be found even if it's the second battery in the chain.

Addressing anything else (0x10 per docs for the 'master' battery, or 0x00) does not work.

I'm also getting random alarms from my MPPT (also Victron) with an error code 67 -> Lost connection to the BMS.  Once DVCC is on the MPPT works in coordination as I understand it.  When I get the error, I end up needing to reset it out of networked comms and tell my Multiplus to reset the VE bus, to pull it out of passthrough mode.

I'm continuing to look at this.  Getting ready to hook up a windows laptop and see what the tool sends/receives to see if it sparks and idea.  If I can send anything for you to look at, let me know.

**Comment by tuxntoast** [comment]
I had read that document when I was trying to write this module, however it did not seem accurate. There was another guide I had found that was more inline with the hardware I had, but still was not 100% accurate, so there was a lot of reverse engineering that I had to do using the client software / testing. I believe I had a comment in my code with the documentation url that I had found to be the closest to what I was seeing on my BMS. Not sure if that was just because I was using the 12v version… calling the vendor the bMS on the 12v is old, and has not been updated in a long time. 24v/ and 48v have newer bMS software, and might be more inline with the link you attached to your message.

**Comment by tuxntoast** [comment]
How many batteries do you have? What is their voltage? I currently only have one 12v 400ah pack, and reached the limits of my rig testing this software. If you also have 12v and more then one, you might be able to assist with getting me the client logs I would need to understand what would be needed to fixed. 

As for wiring, USB -> RS485 to the left port with the BMS at x01address. I believe there is a battery ID in the command that is being sent (python variable command_get_version, command_get_stats, command_get_config). But only having a single BMS I was not able to test and confirm this. There might be some edits that are needed if you want to address more than on BMS. Would need logs from the BMS communications client eg4 bms tools to better understand what is needed. 

For Error 67, watch your RS485 cable. I had issues with electrical interference from the battery leads. Routing cable via different path help address that error being dropped. DVCC is tied to BMS communication. If the communication is lost, the Max current charge / discharge will be limited until communications is restored. I have not tried putting ferrite beads on the RS485 yet, that also might help.

**Reply by epierce78** [reply]
I have 2 of the 12Vx400Ah batteries.  I was able to get comms working on one battery with hex address 0x10 after looking at the debug data coming out of the windows app for looking at the BMS.  I'm going to attach that log file.  There's a lot of no-ops in this as I let it just 'watch' the battery for a bit.  I changed this code (you should be able to search the attached files for the hex codes):

```
        # self.command_get_version = address + b"\x03\x00\x69\x00\x23\xD4\x0F"
        # self.command_get_stats = address + b"\x03\x00\x00\x00\x27\x05\xD0"
        # self.command_get_config = address + b"\x03\x00\x2D\x00\x5B\x94\x38"

        self.command_get_version = address + b"\x03\x00\x69\x00\x17\xD6\x99"
        self.command_get_stats = address + b"\x03\x00\x00\x00\x27\x06\x91"
        self.command_get_config = address + b"\x03\x00\x2D\x00\x5B\x97\x79"
```

The commented out lines work only with modbus address 0x01 (all based on the branch @mr-manuel sent me, so that initial address is being passed in by the driver).

The non-commented lines work with address 0x10 (BMS master), all DIP switches 'ON' (down).  I haven't had time to sit and noodle those differences much.  It's clear it's probably doing something with the CRC bits, but ... the couple minutes I spent on it didn't show me an easy pattern.  The other EG4 Lifepower driver docs have some info on the CRC calc for those, but don't know if that applies here.

The log file I'm attaching did not have the second battery hooked up with the RS485 comms cable, so if you need another set of logs, I'm happy to do that, probably tomorrow.

[eg4ll_2batt_commlog.txt](https://github.com/user-attachments/files/16117377/eg4ll_2batt_commlog.txt)

**Reply by Justice12t** [reply]
I've got 3x EG4 12v connected.  Let me know how I can help.  I've been running into the same issue. I can get 1 battery to display but that's it.

**Comment by tuxntoast** [comment]
Thanks for the logs. Yep, it looks like there is just a difference in the address that the client is calling to query info. x01for address 1 and x10 for master. I was thinking the last 2 bits had been CRC, but you commands also are different on the 3rd to the last bit as well. When you get a reporting back from the driver, after changing this command, are you able to see the voltage for all 8 cells ( 4 cells from BMS 1 and 4 cell from BMS 2). If so, that would imply that the master is doing a rollup of the slaves that are connected. If not, it would mean that code would need to be written to query all the ID, and then avg out their current values, and roll that up to the Victron OS

**Reply by mr-manuel** [reply]
If you need help write to `support AT eg4electronics.com`, they will give you all the informations needed. I was already in contact with them to get the protocol PDF's.

**Reply by epierce78** [reply]
I pulled some additional logs this morning, that should help us.  Unfortunately, I think the BMS tool is fairly indiscriminate about the commands it sends, as it just fires off one for each possible battery and waits for a timeout.  Within the tool I can only really look at one battery in detail at a time.

With the branch I have on my Cerbo, which includes that change to query the 0x10 address, I still only see the single battery, even though they are both hooked up.  0x10 has the USB/RS485 cable -> Cerbo, then the supplied cable from 0x10 to 0x01.

I'll paste the snippet of all the possible commands below.  The question is (IMO) is it worth trying to reverse engineer the last 3  values, or just set a suffix depending on the ID of the batteries that are being queried?

Here's the output.  I skipped 0x01 since we already know it, and it does work and return that pack's information.

```
\par \cf1 COM4 Send : 02 03 00 69 00 17 D5 EB 
\par \cf3 COM4:Receive<<     Pack 2:State_Query_Overtime
\par \cf1 COM4 Send : 03 03 00 69 00 17 D4 3A 
\par \cf3 COM4:Receive<<     Pack 3:State_Query_Overtime
\par \cf1 COM4 Send : 04 03 00 69 00 17 D5 8D 
\par \cf3 COM4:Receive<<     Pack 4:State_Query_Overtime
\par \cf1 COM4 Send : 05 03 00 69 00 17 D4 5C 
\par \cf3 COM4:Receive<<     Pack 5:State_Query_Overtime
\par \cf1 COM4 Send : 06 03 00 69 00 17 D4 6F 
\par \cf3 COM4:Receive<<     Pack 6:State_Query_Overtime
\par \cf1 COM4 Send : 07 03 00 69 00 17 D5 BE 
\par \cf3 COM4:Receive<<     Pack 7:State_Query_Overtime
\par \cf1 COM4 Send : 08 03 00 69 00 17 D5 41 
\par \cf3 COM4:Receive<<     Pack 8:State_Query_Overtime
\par \cf1 COM4 Send : 09 03 00 69 00 17 D4 90 
\par \cf3 COM4:Receive<<     Pack 9:State_Query_Overtime
\par \cf1 COM4 Send : 0A 03 00 69 00 17 D4 A3 
\par \cf3 COM4:Receive<<     Pack 10:State_Query_Overtime
\par \cf1 COM4 Send : 0B 03 00 69 00 17 D5 72 
\par \cf3 COM4:Receive<<     Pack 11:State_Query_Overtime
\par \cf1 COM4 Send : 0C 03 00 69 00 17 D4 C5 
\par \cf3 COM4:Receive<<     Pack 12:State_Query_Overtime
\par \cf1 COM4 Send : 0D 03 00 69 00 17 D5 14 
\par \cf3 COM4:Receive<<     Pack 13:State_Query_Overtime
\par \cf1 COM4 Send : 0E 03 00 69 00 17 D5 27 
\par \cf3 COM4:Receive<<     Pack 14:State_Query_Overtime
\par \cf1 COM4 Send : 0F 03 00 69 00 17 D4 F6 
\par \cf3 COM4:Receive<<     Pack 15:State_Query_Overtime
\par \cf1 COM4 Send : 10 03 00 69 00 17 D6 99 
```

Also realized I need to hook up the laptop again and query the actual parameters, since I believe those vary per pack.  I'm going to make a rough change to the code.

**Comment by tuxntoast** [comment]
My understanding is the Cerbo is only able to report data on a single battery. So in ay situations that where there is more than one BMS, the driver would need to poll each battery, and rollup their data into a single set of values for display on the touch screen / ui. 

How many cell are reported in the Cerbo with your changes and looking to the battery of an ID of 0x10? Are you still only reading out 4 cells (the 4 cells from the master address), or are you able to see the cells from battery with ID of 01 as well (and thus a total of 8 cells reported)? 

To your question of effort on the CRC…. Either could be done. Ideally the driver should be able to query the master, and know how many BMS are chained together for its power source. It should then be able to poll each of those BMS’s and get their current stats. Code would be shorter and cleaner if we had an understand of how to generate the CRC 2 bytes at the end of the command, but not a requirement. Just would be more variables that would have to be managed, and breaking a few things out to a function to help limit a bunch of copy paste logic. If the master BMS is not aware of all the others units it is connected to, then there would have to be a config that a user is setting at install that denotes how many BMS are connected, and thus we would know how many address to call.

**Reply by epierce78** [reply]
I modified the code I have to work with batteries at modbus address 0x10 and 0x01.  It sees both batteries without issue.  I installed BatteryAggregator, and it's able to pull that data together.

![image](https://github.com/mr-manuel/venus-os_dbus-serialbattery/assets/3248671/89318cce-79f8-42d0-b3de-93737b269b81)

Each battery shows its own data, although I put in a small change for the unique identifier.  For some reason, the serial number of my batteries is ... the same?  So I put the modbus address at the end of the serial number for now.

![image](https://github.com/mr-manuel/venus-os_dbus-serialbattery/assets/3248671/cb110178-1c57-4575-a060-be2c8caec79d)
![image](https://github.com/mr-manuel/venus-os_dbus-serialbattery/assets/3248671/7e2d86e2-a96d-4b7d-a5fe-7d4e5a43852f)

The system is set with BatteryAggregator as the BMS.  I'm going to run a discharge tonight.

Let me know what I can provide/do to assist.

<edit>

The aggregator seems to be doubling the amps being pulled.  I haven't looked at that addon in detail -- it might be predicated on the idea that each device is on its own port.

![image](https://github.com/mr-manuel/venus-os_dbus-serialbattery/assets/3248671/1d9c8456-a85f-49c9-8243-03fd2f57ce48)

My SmartShunt/BMV is showing the correct draw.

**Comment by tuxntoast** [comment]
You have a RS485 connection to both of those different batteries? That is one way manage a few of these. Little more unwieldy when you get to 16.  

The screen shots you included with the mod address is actually the port that each BMS is connected to the Cerbo. Serial is the protocol that is connected to the linux kernel at /dev/ttyUSB0_XXX. I am pulling the serial number from the firmware, but it would be printed out a little lower than what is shown in the screenshot.

I was just looking around the signature solar web site looks like there is a Gen2 unit out on the 12v 400ah units. Was able to find a Gen1 units referbished unit at a discount. Should be here in a few weeks, and should enable me to have the kit needed to wrap up multiple BMS management, and config reading.  I had hopes that some of that extra wiring could be taken up by communication from the BMS protocol. I should find out soon!

**Reply by epierce78** [reply]
So clarification:  I have 2 batteries, 1 RS485 connection from 0x10 to the Cerbo.  The second battery is connected to the first via the supplied communication cable.

I was curious if BatteryAggregator would work, and it does not report the correct values (they're doubled).

**Comment by tuxntoast** [comment]
I have not used battery aggregator but I did look at it a while ago. My understanding was that it would need a usb cable to each of the batteries, and a different port for each. It would then connect and poll each of them independently then roll up the values to the cerbo for display. The screen shots that you had shared showed that the batter violates was different on both of those units, if everything is doubled how did the cerbo see different voltage levels of each of the battery packs?Sent from my iPhoneOn Jul 9, 2024, at 4:52 AM, Eric Pierce ***@***.***> wrote:﻿
So clarification:  I have 2 batteries, 1 RS485 connection from 0x10 to the Cerbo.  The second battery is connected to the first via the supplied communication cable.
I was curious if BatteryAggregator would work, and it does not report the correct values (they're doubled).

**Comment by Justice12t** [comment]
2024-07-09 21:07:31.273534500 INFO:SerialBattery:== Modbus packet good ==
2024-07-09 21:07:31.275046500 INFO:SerialBattery:Connection established to EG4_LL
2024-07-09 21:07:31.279171500 INFO:SerialBattery:dbusname_post = com.victronenergy.battery.ttyUSB2__01
2024-07-09 21:07:36.304453500 INFO:SerialBattery:Found existing battery with DeviceInstance = 1
2024-07-09 21:07:36.638478500 INFO:SerialBattery:DeviceInstance = 1
2024-07-09 21:07:36.640610500 ERROR:SerialBattery:** DRIVER STOPPED! Another battery with the same serial number/unique            identifier "2022-10-25" found! **
2024-07-09 21:07:36.641029500 ERROR:SerialBattery:Please check that the batteries have unique identifiers.
2024-07-09 21:07:36.641734500 ERROR:SerialBattery:Change the customizable field in your BMS settings to be unique.
2024-07-09 21:07:36.642339500 ERROR:SerialBattery:To see which battery already uses this serial number/unique identifier            check this file "/var/tmp/dbus-serialbattery_1.pid"
2024-07-09 21:41:08.585319500 *** CCGX booted (0) ***


This is what I'm getting with 0x10, 0x01 set as the addresses.  Is this what you were seeing before?

**Reply by Justice12t** [reply]
I have 3 batteries connected with different IDs set via the dip switches.  Similar to above, the batteries are connected together via the cables provided when purchased and then via 485 to usb.

**Reply by mr-manuel** [reply]
You can set this to true:

https://github.com/mr-manuel/venus-os_dbus-serialbattery/blob/53d930fe60e85c933df917af179e3817c69c6cc2/etc/dbus-serialbattery/config.default.ini#L417-L421

**Comment by tuxntoast** [comment]
Not clear are those batteries chained together with one rs485 to the cerbo, or is each battery have a cable connected to a usb port. Reading your logs it seems like you might have more than one device plugged into your usb ports. Sent from my iPhoneOn Jul 9, 2024, at 4:12 PM, Justice12t ***@***.***> wrote:﻿
I have 3 batteries connected with different IDs set via the dip switches.  Similar to above, the batteries are connected together via the cables provided when purchased and then via 485 to usb.

**Comment by tuxntoast** [comment]
I am working on testing v1.3.20240705, but I am seeing a common issue get tripped and not sure what might be the cause. 

The driver will run just fine, but when it is charging, it reaches an error state #8 Internal Calculation error after some time as it gets near the max battery / max cell voltage. The driver will run fine when the solar panels are dark, but if it is at 100% charge, and just a little low on the cell voltage, it will trip shortly after first light on the panels. 

Loading the same driver on a prior version of the bus-serial-battery does not trigger this. Reading over the code it seems there is a bunch of new code that has been added this this project around better managing cell voltages and backing off from charge voltage to float. Do you have some insight of what be be the cause of this Internal Calculation error?

**Reply by mr-manuel** [reply]
Checking the docs and using the search is always a good idea: https://mr-manuel.github.io/venus-os_dbus-serialbattery_docs/troubleshoot/#error-codes

**Reply by mr-manuel** [reply]
With older driver versions the errors where not shown in the GUI and therefore nearly no one knew that there were errors in the logs. This error was introduced to let the user know that there is something faulty. Check the logs and you will know what is going on.

**Comment by tuxntoast** [comment]
Got my second BMS unite, and able to make some changes to add support for multi-bms data collection / reporting. 

Still testing out stability but functionality see to mostly be there.   I did hit a limit when pushing cell stats to the Cerbo. Pushing more then 4 cells, means that the DVCC Voltage is going to be calculated across ALL 8 cells(28v vs 14). To get something working, I am only pushing the cell voltage of the first BMS, but am doing roll ups / averages / summaries  of all of the BMS in the communication chain and the collected data points. 

I am testing only on a 12v pack, and think that other voltage packs should be supported by this, but would need some one to aid in testing that.  I also only have 2 BMS units, but have written to support all 16 ID's that could be chained together. 

You can check out this code, and dump that into `/data/etc/dbus-serialbatter/bms/eg4_ll.py`
https://github.com/tuxntoast/eg4-ll/blob/main/egll.py

`run /data/etc/dbus-serialbattery/reinstall-local.sh`
and restart the process pointing to USB0
` /data/etc/dbus-serialbattery/start-serialbattery.sh ttyUSB0 &`
tailing logs via: 
`tail -F -n 1000 /data/log/dbus-serialbattery.ttyUSB0/current | tai64nlocal`

There is a setting of the ID your BMS are set to. This helps speed up discovery / connection. Make sure to update the value of `batteryPackId` on line 66

Setting the value of statuslogger on line 61, prints detailed per BMS Units stats. Can be used to look over what is being collected, and what is being reported to the cerbo. Great for testing. 

Wiring: 
Just need a single RS485-> USB cable between the Cerbo and the master bms (id 16) 
then a Cat5 Cable between the RS485 ports on each of the BMS units. No need to have each BMS directly connected to the Cerbo via a USB cable

**Reply by mr-manuel** [reply]
How is it going with your changes? Do you think that we can integrate that in the main driver? I saw that the code is quite different comparing to the different one.

If adding the daisy chain feature for this BMS, do we also need to add some explanation to this page? https://mr-manuel.github.io/venus-os_dbus-serialbattery_docs/next/general/connect

**Comment by tuxntoast** [comment]
I did do a major refactor, to add support for more then one bms units, as well as support BMS id’s from 1 to 64. That is running good. There is one more area I still need to work on. And that would be pushing cell details from any other bms that might be in the communication chain other then the master. Right now the data is summarized and pushed to the cerbo, but you can’t see the cell voltage of the other units in the chain (other then tailing the logs) If you are wanting to get a release out, I could hold back these change and we could push the other parts that I did get re-written

Sent from my iPhoneOn Oct 7, 2024, at 11:26 PM, Manuel ***@***.***> wrote:﻿
How is it going with your changes? Do you think that we can integrate that in the main driver? I saw that the code is quite different comparing to the different one.

**Reply by mr-manuel** [reply]
Ok, I was only curious :-) why do you want to summarize the batteries instead of displaying every battery in the chain as single battery like the other BMS drivers do?

**Comment by tuxntoast** [comment]
These batteries can be chained up to 64 units. Serial communication is very temperamental. And if I keep the driver to only deal with one driver, the way dubs-serial battery works would be attempting to create a new serial connection to the master bms each polling cycle. Doing this on a large array will result in a lot of no replies. I ended up having to pull into my driver the code for creating the serial connection, as well as the function to send and get the reply in a graceful manor to how these bms’s operate.Sent from my iPhoneOn Oct 8, 2024, at 6:57 AM, Manuel ***@***.***> wrote:﻿
Ok, I was only curious :-) why do you want to summarize the batteries instead of displaying every battery in the chain as single battery like the other BMS drivers do?
